src/ua.c

```diff
@@ -1323,6 +1366,47 @@ static void sipsess_conn_handler(const struct sip_msg *msg, void *arg)
 }
 
 
+static void ua_xhdr_filter_destructor(void *arg)
+{
+	struct ua_xhdr_filter *filter = arg;
+	mem_deref(filter->hdr_name);
+}
+
+
+int ua_add_xhdr_filter(struct ua *ua, const char *hdr_name)
+{
+	struct ua_xhdr_filter *filter;
+
+	if (!ua)
+		return EINVAL;
+
+	if (!ua->hdr_filter) {
+		ua->hdr_filter = mem_zalloc(sizeof(*ua->hdr_filter),
+			(mem_destroy_h *)list_flush);
+
+		if (!ua->hdr_filter) {
+			return ENOMEM;
+		}
+		list_init(ua->hdr_filter);
+	}
+
+	char *buf;
+	if (str_dup(&buf, hdr_name))
+		return ENOMEM;
```

Comment:
instead of using a temp buf, you can instead dup the string
directly into `filter->hdr_name`

in case of errors, the destructor will deref the string.